File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        answer = set()
        nums.sort()
        target = 0
        for i in range(len(nums)):
            lo = i + 1
            hi = len(nums) -1
            a = nums[i]
            while lo < hi:
                if a + nums[lo] + nums[hi] < target:
                    # too small, increase sum
                    lo += 1
                elif a + nums[lo] + nums[hi] > target:
                    # too large, decrase sum
                    hi -= 1
                else:
                    # equals target
                    answer.add(tuple([a, nums[lo], nums[hi]]))
                    lo += 1
                    hi -= 1
        return answer
    
# Time: O(N^2)
# Space: O(N)

File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let answer = [];
    nums.sort((a,b)=>a-b)
    for (let i=0; i < nums.length; i++) {
        if (i>0 && nums[i-1] == nums[i]) {
            continue
        }
        let a = nums[i]
        let lo = i + 1
        let hi = nums.length - 1
        while (lo < hi) {
            if (a + nums[lo] + nums[hi] > 0) {
                hi--;
            } else if (a + nums[lo] + nums[hi] < 0) {
                lo++;
            } else {
                answer.push([a, nums[lo], nums[hi]])
                lo++;
                hi--;
                while (lo < hi && nums[lo] == nums[lo-1]) {
                    lo++;
                }
                
            }
            
        }
    }
    return answer
    
};